pkg/function/service.go

```diff
@@ -81,26 +117,47 @@ func (fs *Service) MapFn(ctx context.Context, d *functionpb.Datum) (*functionpb.
 // ReduceFn applies a reduce function to a datum stream.
 func (fs *Service) ReduceFn(stream functionpb.UserDefinedFunction_ReduceFnServer) error {
 	var (
-		ctx      = stream.Context()
-		key      string
-		reduceCh = make(chan Datum)
-		md       Metadata
+		key       string
+		md        Metadata
+		err       error
+		startTime int
+		endTime   int
+		reduceCh  = make(chan Datum)
+		ctx       = stream.Context()
 	)
 
-	// get key and metadata from gPRC metadata
-	if grpcMD, ok := metadata.FromIncomingContext(ctx); ok {
-		// get Key
-		keyValue := grpcMD.Get(DatumKey)
-		if len(keyValue) > 1 {
-			return fmt.Errorf("expect extact one key but got %d keys", len(keyValue))
-		} else if len(keyValue) == 1 {
-			key = keyValue[0]
-		} else {
-			// do nothing: the length equals zero is valid, meaning the key is an empty string ""
-		}
-		// TODO: get metadata
+	grpcMD, ok := grpcmd.FromIncomingContext(ctx)
+	if !ok {
+		return fmt.Errorf("key and window information are not passed in grpc metadata")
+	}
+
+	// get key from gPRC metadata
+	key, err = getValueForKey(grpcMD, DatumKey)
+	if err != nil {
+		return err
+	}
+
+	// get window start and end time from grpc metadata
+	var st, et string
+	st, err = getValueForKey(grpcMD, WinStartTime)
+	if err != nil {
+		return err
+	}
+
+	et, err = getValueForKey(grpcMD, WinEndTime)
+	if err != nil {
+		return err
 	}
 
+	startTime, _ = strconv.Atoi(st)
```

Comment:
should use strconv.ParseInt functions here for int64 values.

Reply:
yes, done